features/lorawan/lorastack/phy/LoRaPHY.cpp

```diff
@@ -123,316 +106,1317 @@ void LoRaPHY::handle_send(uint8_t *buf, uint8_t size)
+            get_phy.datarate = dr_out;
+            phy_param = get_phy_params(&get_phy);
+            dr_out = phy_param.Value;
+
+            if (dr_out == phy_params.min_tx_datarate) {
+                // We must set adrAckReq to false as soon as we reach the lowest datarate
+                set_adr_ack_bit = false;
+                if (restore_channel_mask) {
+                    // Re-enable default channels
+                    restore_default_channels();
+                }
+            }
+        }
+    }
+
+    return set_adr_ack_bit;
+}
+
+void LoRaPHY::compute_rx_win_params(int8_t datarate, uint8_t min_rx_symbols,
+                                    uint32_t rx_error,
+                                    rx_config_params_t *rx_conf_params)
+{
+    double t_symbol = 0.0;
+
+    // Get the datarate, perform a boundary check
+    rx_conf_params->datarate = MIN( datarate, phy_params.max_rx_datarate);
+
+    rx_conf_params->bandwidth = get_bandwidth(rx_conf_params->datarate);
+
+    if (phy_params.fsk_supported && rx_conf_params->datarate == phy_params.max_rx_datarate) {
+        // FSK
+        t_symbol = compute_symb_timeout_fsk(((uint8_t *)phy_params.datarates.table)[rx_conf_params->datarate]);
+    } else {
+        // LoRa
+        t_symbol = compute_symb_timeout_lora(((uint8_t *)phy_params.datarates.table)[rx_conf_params->datarate],
+                                            ((uint32_t *)phy_params.bandwidths.table)[rx_conf_params->datarate]);
+    }
+
+    get_rx_window_params(t_symbol, min_rx_symbols, rx_error, RADIO_WAKEUP_TIME,
+                         &rx_conf_params->window_timeout, &rx_conf_params->window_offset);
+}
+
+bool LoRaPHY::rx_config(rx_config_params_t* rx_conf, int8_t* datarate)
+{
+    radio_modems_t modem;
+    uint8_t dr = rx_conf->datarate;
+    uint8_t max_payload = 0;
+    uint8_t phy_dr = 0;
+    uint32_t frequency = rx_conf->frequency;
+
+    _radio->lock();
+
+    if (_radio->get_status() != RF_IDLE) {
+        _radio->unlock();
+        return false;
+    }
+
+    _radio->unlock();
+
+    if (rx_conf->rx_slot == RX_SLOT_WIN_1) {
+        // Apply window 1 frequency
+        frequency = phy_params.channels.channel_list[rx_conf->channel].frequency;
+        // Apply the alternative RX 1 window frequency, if it is available
+        if (phy_params.channels.channel_list[rx_conf->channel].rx1_frequency != 0) {
+            frequency = phy_params.channels.channel_list[rx_conf->channel].rx1_frequency;
+        }
+    }
+
+    // Read the physical datarate from the datarates table
+    uint8_t *datarate_table = (uint8_t *) phy_params.datarates.table;
+    uint8_t *payload_table = (uint8_t *) phy_params.payloads.table;
+    uint8_t *payload_with_repeater_table = (uint8_t *) phy_params.payloads_with_repeater.table;
+
+    phy_dr = datarate_table[dr];
+
+    _radio->lock();
+
+    _radio->set_channel(frequency);
+
+    // Radio configuration
+    if (phy_dr == DR_7 && phy_params.fsk_supported) {
```

Comment:
FSK should be used if dr == DR7:
if (phy_dr == DR_7 && phy_params.fsk_supported) {

Reply:
Hmmm, This is weird if it would cause excessive retransmissions. Because dr is the index in datarate table whereas phy_dr is the value of data rate. In EU, they are actually synonymous, so DR_0 resides at index 0.

Reply:
Not exactly. In EU case table is for example: 
static const uint8_t datarates_EU868[]  = {12, 11, 10,  9,  8,  7,  7, 50};